Convert a Drupal 8 site to a Drupal 9 site managed with Integrated Composer (#6145)
pantheon-systems/documentation: source/content/guides/composer-convert.md

### Contributed Code

#### Modules and Themes

The goal of this process is to have Composer manage all the site's contrib modules, contrib themes, core upgrades, and libraries (we'll call this _contributed code_). The only things that should be migrated from the existing site are custom code, custom themes, and custom modules that are specific to the existing site.

The steps here ensure that any modules and themes from [drupal.org](drupal.org) are in the `composer.json` `require` list.

Once Composer is aware of all the contributed code, you'll be able to run `composer upgrade` from within the directory to have Composer upgrade all the contributed code automatically.

Begin by reviewing the existing site's code. Check for contributed modules in `/modules`, `/modules/contrib`, `/sites/all/modules`, and `/sites/all/modules/contrib`.

1. When reviewing the site, take stock of exactly what versions of modules and themes you depend on. One way to do this is to change to run the `pm:projectinfo` Drush command from within a contributed modules folder (e.g. `/modules`, `/themes`, `/themes/contrib`, `/sites/all/themes`, `/sites/all/themes/contrib`, etc.).

>   This command works on Drush 8. If you're using Drush 9, use `pm:list` or refer to [Drush Commands](https://drushcommands.com/drush-9x/pm/pm:projectinfo/):

Reviewer: Users with Drush 9 are probably on Composer-managed sites, and should be following a different guide. n.b. pm:list shows modules, and pm:projectinfo shows projects. Projects are collections of modules. If the user listed modules, they'd need to figure out the project name from the module name. This is usually the same for projects with one module. Projects with multiple modules often have one with the same name as the project.

Reviewer: There is no point in linking to the Drush 9 pm:projectinfo entry on Drush Commands, though; if you run pm:projectinfo you just get an error message telling you to use pm:list or composer show

As of Drush 8.4.8 and Drush 9.7.3, though, you can run `pm:list --fields=*` and see all of the available modules along with the project they're part of. This field isn't included in Drush 10.4.0, but it will be in the next release (hopefully available soon)

  ```bash{promptUser:user}
  terminus drush $SITE.dev -- pm:projectinfo --fields=name,version --format=table
  ```

  This will list each module followed by the version of that module that is installed.

1. You can add these modules to your new codebase using Composer by running the following for each module in the `$SITE-composer` directory:

  ```bash{promptUser:user}
  composer require drupal/MODULE_NAME:^VERSION
  ```

>   Where `MODULE_NAME` is the machine name of the module in question, and `VERSION` is the version of that module the site is currently using. Composer may pull in a newer version than what you specify, depending upon what versions are available. You can read more about the caret (`^`) in the [Composer documentation](https://getcomposer.org/doc/articles/versions.md#caret-version-range-).

Reviewer: At one point we had an explanation about versions strings.

Old-style Drupal version strings:
```
Chaos Tools (ctools)  8.x-3.4
```
Drop the `8.x-` to convert this into `^3.4`

Semver version strings:
```
Devel (devel)  4.1.1
```
Use this version directly, e.g. `^4.1.1`

Author: I've added this a subpoints in https://github.com/pantheon-systems/documentation/pull/6145/files#diff-de5e95860ba4c809f9e43c75a3ff0cd70ad67998ae04beec2df43d8d52958415R200-R216

   - If you get the following error, the module listed in the error (or its dependencies) does not meet compatibility requirements:

   ```none
   [InvalidArgumentException]
   Could not find a version of package drupal/MODULE_NAME matching your minimum-stability (stable). Require it with an explicit version constraint allowing its desired stability.
   ```

>    If there is no stable version you can switch to, you may need to adjust the `minimum-stability` setting of `composer.json` to a more relaxed value, such as `beta`, `alpha`, or even `dev`. You can read more about `minimum-stability` in the [Composer documentation](https://getcomposer.org/doc/04-schema.md#minimum-stability).

Reviewer: We should not recommend lowering the minimum stability to `dev`, as this can have odd side effects. We should instead tell people to require the specific version they need via Composer, e.g.
```
composer require drupal/ctools:^3.4@dev
```
If requiring a dev version fails because the dev version of the module requires a dev version of a dependency, they should fix this by also allowlisting the dev dependency at the same time:
```
composer require drupal/some-module:^1@dev org/some-dependency:^2@dev
```

**Trust a robot?**

One of Pantheon's engineers got tired of doing this process by hand, so he trained some robots to identify modules and put them into `composer.json`.

Robots are cool, but they're not perfect, so you should understand the goal of this process as well as the limitations of automating the process.

<Accordion title="A script that can help add modules to composer.json" id="modules-script" icon="cogs">

First, disclaimers:

> Pantheon support is not available for this script or anything that results from trying to use it.

Reviewer: I think we need slightly higher levels of support for scripts we publish in official documentation. This statement sounds a little too strong. Do we have precedence for this (support or lack of support for scripts in documentation)?

I think we should publish the docs without the scripts first & then have a discussion about support. e.g. with nominal buy-in from Devrel we could perhaps direct folks to #composer-workflow

Reviewer: Regarding precedence, we have this text in Terminus plugins:

> This repository is provided without warranty or direct support. Issues and questions may be filed in GitHub but their resolution is not guaranteed.

  ```

  If you prefer, you can replace `composerify` with another branch name. If you do, remember to adjust the other examples in this doc to match.

1. Copy any existing configuration from the default branch. If no files are copied through this step, that's ok:

  ```bash{promptUser:user}
  git checkout master sites/default/config
  git mv sites/default/config/* config
  git rm -f sites/default/config/.htaccess
  git commit -m "Pull in configuration from default branch"
  ```

1. Check for `pantheon.yml` settings you need to preserve by comparing your old codebase's `pantheon.yml` to the new `pantheon.upstream.yml`:

  ```bash{promptUser:user}
  git diff master:pantheon.yml pantheon.upstream.yml
  ```

  Press `q` on your keyboard to exit the diff display.

   - If there are settings from `pantheon.yml` (shown with a `-` in the diff output), consider copying over your old `pantheon.yml` to preserve these settings:

     ```bash{promptUser:user}
     git checkout master pantheon.yml
     git add pantheon.yml
     git commit -m 'Copy my pantheon.yml'
     ```

>    If you prefer to keep the value for `database` from `pantheon.upstream.yml`, remove it from `pantheon.yml`.

Reviewer: :rotating_light: :rotating_light: :rotating_light:  DANGER :rotating_light: :rotating_light: :rotating_light:
If a user has an existing site that's running Mariadb 10.0 and you follow this guide and end up with a pantheon.yml or pantheon.upstream.yml with a database version of 10.4 before Drupal 9 General Availability, your site might be in danger of database corruption.

Reviewer 2: Currently trying to this on an actual project. I set 
```yml
api_version: 1
database:
  version: 10.0
```

on my `pantheon.yml` to prevent using 10.4

but I got 

```
remote: 
remote: PANTHEON ERROR:
remote: 
remote: Changes to `pantheon.yml` detected, but there was an error while processing it:
remote: 
remote: pantheon.yml: 
remote: Validation failed with error:
remote: >   10.0 is not one of [10.4]
remote: 
remote: while validating the following value:
remote: >   10.0
remote: >   ...
remote: 
```

Reviewer: @hanoii The `database:` directive in pantheon.yml is not ready to use, and is dangerous. You must remove the directive if you do not want to use it, because at the moment the only valid value is 10.4. If your pantheon.upstream.yml has the `database:` directive, though, then you must remove it there as well; failure to do so prior to GA could be dangerous. An individual site is not supposed to modify the pantheon.upstream.yml file, though, so you will have extra work to do later if you do this. I would like to reiterate that if you make a mistake during this process, the result could be dangerous. If you think that you may have applied a `database` directive with the value 10.4 in any existing site's multidev, I would recommend deleting that multidev, and start over with a fresh one.

The recommended process to follow today is https://github.com/pantheon-systems/documentation/pull/6314. I would follow this guide to convert an existing drops-8 to integrated composer, even if your intention is to downgrade back to Drupal 8.

Reviewer 2: @greg-1-anderson thanks! So no `database:` directive on either `pantheon.yml` and `pantheon.upstream.yml` means `10.0`? Isn't there anyway for me to override the `database` on my `pantheon.yml` with null or empty object something that would mean the same as no database? 

Anyway, I will try removing it. Do you foresee any issue moving forward that makes me reconsider doing the d8 to a new d9 site worthwhile considering at this point? I assume that  changing the upstream yaml, and then reverting that commit should allow for proper upstream merging from now on once 10.4 is GA.

Reviewer: There is a comment above from mbaynton that I forgot about: it appears that this you can get 10.0 by modifying only pantheon.yml as follows:
```
database:
  {}
```
My honest advice, though, is that you should follow the instructions to convert your site to Drupal 9 by creating a new site, and then move your work back to your existing site after D9 GA.

Reviewer 2: @greg-1-anderson ok. If I move it to a new d9 site I guess I can simply move everything there and then switch domains. Will explore this option. Thanks!

Reviewer 2: FWIW, `database: {}` does work in that I get the push to go through and the site/code looks to have built, howeveer I connected to the mysql site and it reported to be `10.4`. The site worked, but I am not trying without any database at all.

 I did find a few things odd. The `tag pantheon_composerify_1` actually has the code from `master` (not yet composerified), but the deployed code as seen by SFTP looks the new stuff. This seems to be some odd issue with multidev and IC. I also saw the same behavior on a new d9 site I am using for testing where I created a new multidev. There it is less obvious because the tag belongs to the last commit, but there seems to be no artifacts tag on multidevs.

Reviewer 3: Here's mine from my last project:
```yaml
api_version: 1
web_docroot: true
php_version: 7.4
drush_version: 10
database:
  version: 10.4
```

Reviewer 3: FYI: Decoupled react with D9 backend: https://live-freshdrupalmi.pantheonsite.io

type: guide
permalink: docs/guides/:basename
cms: "Drupal 8"
categories: [develop]
tags: [composer, site, workflow]
contributors: [dustinleblanc, greg-1-anderson]
reviewed: "2020-12-01"
---

Drupal sites often require the use of Composer to manage site dependencies. The need to begin using Composer for a site build can often surface after a site is in development, necessitating a conversion of the site's codebase. At a high level, the goals of the conversion are to remove dependencies that Composer will manage from your Git repository, and tell Composer about those dependencies instead.

Existing sites that wish to upgrade from Drupal 8 to Drupal 9 and convert to Integrated Composer, or just convert to Integrated Composer and defer the Drupal 9 upgrade will need to go through some manual steps.

## Before You Begin

- Review our documentation on [Git](/git), [Composer](/composer), and [Terminus](/terminus), and have them installed and configured on your local computer.
   - Mac users can use [Homebrew](https://brew.sh/) to install both Git and Composer, along with their required dependencies:

     ```bash{promptUser:user}

>      brew install git composer

Reviewer: It might also be a good idea to advise users that they should be on Composer 2. If we do this, then we should probably also link to a document with guidance. If we don't do this, they may run into problems if they push a site that only works with Composer 1 up to the platform, which uses Composer 2.

---
title: Upgrade a Drupal 8 Site to a Composer-Managed Drupal 9 Site
description: Upgrade a Drupal 8 site to integrated Composer and Drupal 9.
type: guide
permalink: docs/guides/:basename
cms: "Drupal"
categories: [develop]
tags: [composer, site, workflow]
contributors: [dustinleblanc, greg-1-anderson]
reviewed: "2020-12-01"
---

Drupal 9 sites on Pantheon have Composer built-in to manage site dependencies.

For a smooth upgrade experience, and to avoid potential conflicts, this doc shows how to migrate a Drupal 8 site to a freshly prepared, new Drupal 9 site.

The goals of this upgrade are to remove dependencies from the old site that Composer will manage from your Git repository, and tell Composer about those dependencies in the new site instead.

Please note, that since you are migrating a site through this process, the new site will no longer maintain your existing commit history.

> ## Before You Begin

Author: added https://github.com/pantheon-systems/documentation/pull/6145/files#diff-de5e95860ba4c809f9e43c75a3ff0cd70ad67998ae04beec2df43d8d52958415R32

---
title: Convert a Standard Drupal 8 Site to a Composer Managed Site

> description: Use Composer to manage Drupal 8 sites often require the use of Composer to manage site dependencies. The need to begin using Composer for a site build can often surface after a site is in development, necessitating a divergence from the Pantheon managed Upstream.

Reviewer 4: Since e.g. a Drupal 8 to 9 upgrade wouldn't necessitate a divergence from our D9 managed upstream
```suggestion
description: Drupal 8 and 9 sites often require the use of Composer to manage site dependencies. The need to begin using Composer for a site build can often surface after a site is in development, necessitating a conversion of the site's codebase.
```

type: guide
permalink: docs/guides/:basename
cms: "Drupal 8"
categories: [develop]
tags: [composer, site, workflow]
contributors: [dustinleblanc, greg-1-anderson]
reviewed: "2020-12-01"
---

> Drupal 8 sites often require the use of Composer to manage site dependencies. The need to begin using Composer for a site build can often surface after a site is in development, necessitating a divergence from the Pantheon managed Upstream.

Reviewer 4: ```suggestion
Drupal sites often require the use of Composer to manage site dependencies. The need to begin using Composer for a site build can often surface after a site is in development, necessitating a conversion of the site's codebase.
```

type: guide
permalink: docs/guides/:basename
cms: "Drupal"
categories: [develop]
tags: [composer, site, workflow]
contributors: [dustinleblanc, greg-1-anderson, stovak]
reviewed: "2021-03-31"
---

Drupal 9 sites on Pantheon have Composer built-in to manage site dependencies.

For a smooth upgrade experience, and to avoid potential conflicts, this guide shows how to migrate a Drupal 8 site to a freshly prepared, new Drupal 9 site.

The goals of this upgrade are to remove dependencies that Composer will manage from the existing site's Git repository, and have Composer manage those dependencies in the new site instead.

Note that since you are migrating a site through this process, the new site will no longer maintain your existing commit history.

<Alert title="Note" type="info">

To maintain best practice and to avoid difficult, time-consuming repairs to the site, this doc is written for users with access to Pantheon's [Multidev](/multidev) feature.

> Pantheon support is not available to users who avoid the Multidev steps.

Reviewer: We can direct people to the "Manually Migrate" page (or the renamed version thereof) to follow the "create a new site" path.

</Alert>

## Steps we'll take

> 1. Convert to Composer

Reviewer 4: Just some extra context about what this guide applies to
```suggestion
1. Convert a Drupal 8 (drops-8) site to Composer
```

1. Upgrade to Drupal 8.9
1. Upgrade to Drupal 9

## Add the Pantheon Integrated Composer Upstream in a New Local Branch

You're about to make massive changes to the codebase. We recommend you to do this work on a new branch, as it might take you some time to complete and rolling back changes can be complicated:

1. In your local terminal, change directories to your site project. For example, if you keep your projects in a folder called `projects` in the home directory:

  ```bash{promptUser:user}
  cd ~/projects/$site/
  ```

1. Add the Pantheon Drupal Upstream as a new remote, fetch the `ic` branch, and checkout to a new local branch based on it called `composerify`:

  ```bash{promptUser:user}
  git remote add ic git@github.com:pantheon-upstreams/drupal-project.git && git fetch ic && git checkout -b composerify ic/master
  ```

>   You can replace `composerify` with another branch name. If you do, remember to adjust the other examples in this doc to match.

Reviewer 4: ```suggestion
  This upstream is a useful starting point for converting Drupal 8 sites to Composer, but it is also a Drupal 9 upstream. The only changes for Drupal 8 are a few version numbers, which we'll edit later.

  You can replace `composerify` with another branch name. If you do, remember to adjust the other examples in this doc to match.
```

1. Copy your configuration from the default branch:

  ```bash{promptUser:user}
  git checkout master sites/default/config

>   git mv sites/default/config/* config

Reviewer 4: ```suggestion
  git mv sites/default/config/* config
  git rm -f sites/default/config/.htaccess
```
The `git mv` glob doesn't move dotfiles, but there's already a config/.htaccess. We end up with an empty `sites/default/config` containing only a .htaccess which I think is unnecessary

  git commit -m "Pull in configuration from default branch"
  ```

> 1. Compare your `pantheon.yml` with the new `pantheon.upstream.yml` and check for conflicts:

Reviewer 4: I didn't understand at first what this step was about, this context about what the goal is could help people to successfully interpret relevant things in the diff they get
```suggestion
1. Check for `pantheon.yml` settings you need to preserve by comparing your old codebase's `pantheon.yml` to the new `pantheon.upstream.yml`:
```

  ```bash{promptUser:user}
  git diff master:pantheon.yml pantheon.upstream.yml
  ```

> 1. If you need your `pantheon.yml`, copy it over (be sure to edit if needed):

Reviewer 4: ```suggestion
1. If there are lines from `pantheon.yml` (shown with a `-` in the diff output), consider copying over your old pantheon.yml to preserve these settings:
```

  ```bash{promptUser:user}
  git checkout master pantheon.yml
  git add pantheon.yml
  git commit -m 'Copy my pantheon.yml'
  ```

> 

Reviewer 4: ```suggestion
Note that a few settings are required to be set as they are in the `pantheon.upstream.yml` for sites using Integrated Composer. These are:
* `web_docroot`
* `build_step`
```
expounding a little on the prior "be sure to edit if needed"

### (Optional) Return to Drupal 8

If you’re not ready to update to Drupal 9 yet. (Advisable: test the Composer conversion before testing the Drupal 9 upgrade)

> Note that you might want to hand-edit your composer.json and remove the drupal/core entry altogether. The empty drupal/core entry from the steps above will cause drupal/core to be erased and redownloaded frequently

Reviewer 4: This sentence confuses me. I don't see `drupal/core` included in the root or upstream-configuration's composer.json, and I am wondering if there's more information available about the problem that was experienced. Composer shouldn't reinstall a dependency unless a newer release is available.

Author: @greg-1-anderson had this in the issue (#6117), and I copied it over - I haven't encountered it either, but I'm just playing with the Umami D8 sample site in my testing.

Reviewer: Remove that part -- I was talking about the PATCH to drupal/core, which is now gone. Sorry for being unclear in my haste.

```bash{promptUser:user}
cd upstream-configuration/

> composer remove zaporylie/composer-drupal-optimizations --no-update

Reviewer 4: ```suggestion
```
No longer needs to be removed because a Composer-2 compatible release was created

composer require drupal/core-recommended:^8 --no-update
cd -

> composer config --json "extra.patches.drupal/core" '{}'

Reviewer 4: ```suggestion
```
No longer needed because the configuration this targets in the upstream was removed in https://github.com/pantheon-upstreams/drupal-project/commit/5ebb134bf1024320b805f4ac2a10d4a7e4908211

> composer update

Reviewer 4: ```suggestion
# Remove new MariaDB directive from pantheon.upstream.yml
sed -i 's/^\s*version: 10.[0-9]*$/  {}/' pantheon.upstream.yml
composer update
```
There's a database version directive in the `pantheon.upstream.yml` that, at present, must be removed for any existing site regardless of Drupal version. I'm including it in this "Optional return to Drupal 8" section though to future-proof this documentation.

Reviewer 4: @EdwardAngert @greg-1-anderson I thought I'd been down this road before. This is how we would want to deal with https://github.com/pantheon-systems/documentation/pull/6145#discussion_r575312499 I believe. Removing the database version will allow us to upgrade the sites as time goes on, so it's better than including the database version and requiring 10.0.

Reviewer 4: Doh, bad idea to modify pantheon.upstream.yml. I've tested however that if you put this into `pantheon.yml` instead, then you override `pantheon.upstream.yml`'s 10.4, and also maintain the desired effect of the database version not being strictly pinned:
```
database:
  {}
 ```
That's what I think I'll have the terminus conversion tool I'm working on do if the user says they're staying with Drupal 8.

Reviewer: I think that this doc is blocked on the database upgrade, so we can just leave this step out.

</Alert>

## Add the Pantheon Integrated Composer Upstream in a New Local Branch

You're about to make massive changes to the codebase. We recommend you to do this work on a new branch, as it might take you some time to complete and rolling back changes can be complicated:

1. In your local terminal, change directories to your site project. For example, if you keep your projects in a folder called `projects` in the home directory:

  ```bash{promptUser:user}
  cd ~/projects/$site/
  ```

> 1. Add the Pantheon Drupal Upstream as a new remote, fetch the `ic` branch, and checkout to a new local branch based on it called `composerify`:

Reviewer: `ic` is a remote, not a branch. The commands below add a new remote called ic, fetches its master branch and then creates a new branch called "composerify" from the master branch of the ic remote.

Note that the resulting branch "composerify" shares a commit history with the drupal-project upstream, and has no commits in common with the site's existing master branch.

  ```bash{promptUser:user}
  git remote add ic git@github.com:pantheon-upstreams/drupal-project.git && git fetch ic && git checkout -b composerify ic/master
  ```

  This upstream is a useful starting point for converting Drupal 8 sites to Composer, but it is a Drupal 9 upstream. To use it for Drupal 8, we will adjust the version numbers in a later step.

  You can replace `composerify` with another branch name. If you do, remember to adjust the other examples in this doc to match.

> 1. Copy your configuration from the default branch:

Reviewer: While it is recommended to store configuration in the database, some sites might not do this. It's not an error if this step does not bring in any files.

  ```bash{promptUser:user}
  git checkout master sites/default/config
  git mv sites/default/config/* config
  git rm -f sites/default/config/.htaccess
  git commit -m "Pull in configuration from default branch"
  ```

> 1. Check for `pantheon.yml` settings you need to preserve by comparing your old codebase's `pantheon.yml` to the new `pantheon.upstream.yml`:

Reviewer: Users should expect their settings to change for `database` and `build_step` and `web_docroot` but they might want to preserve their current settings for `php_version`, `drush_version` & similar

  ```bash{promptUser:user}
  git diff master:pantheon.yml pantheon.upstream.yml
  ```

   - If there are settings from `pantheon.yml` (shown with a `-` in the diff output), consider copying over your old pantheon.yml to preserve these settings:

     ```bash{promptUser:user}
     git checkout master pantheon.yml
     git add pantheon.yml
     git commit -m 'Copy my pantheon.yml'
     ```

   Note that a few settings must be left as they are in `pantheon.upstream.yml` for sites using Integrated Composer. These are:

     * `web_docroot`

>      * `build_step`

Reviewer: Add `database` to this list. Maybe mention that it's best to just remove the setting from pantheon.yml if they want to use the same value as pantheon.upstream.yml.

Author: I tried to explain this as succinctly as possible, but might need to come back with more words if it feels like it's missing the "why"

Reviewer: Using `build_step: true` is the main point of this conversion, so that one should not be overridden.

> ### (Optional) Return to Drupal 8

Author: I've tried to move this to different parts of the doc, but none really feel right, and I'm not sure it belongs here either. Feel free to find a better place for this.

If you have access to the [Multidev](/multidev) feature, test the Composer conversion in a Multidev before testing the Drupal 9 upgrade. If you’re not ready to update to Drupal 9 yet, return to Drupal 8:

```bash{promptUser:user}
cd upstream-configuration/
composer require drupal/core-recommended:^8 --no-update
cd -
# Remove new MariaDB directive from pantheon.upstream.yml
sed -i 's/^\s*version: 10.[0-9]*$/  {}/' pantheon.upstream.yml
composer update
git add .
git commit -m 'Revert to Drupal 8'
```

> ## Create a New Composer Project

Reviewer: This section should be removed (everything down to "## Add in the Custom and Contrib Code Needed to Run Your Site"), as this is part of the old documentation. It is not part of the new procedure.

</Alert>

## Add the Pantheon Integrated Composer Upstream in a New Local Branch

> You're about to make massive changes to the codebase. We recommend you to do this work on a new branch, as it might take you some time to complete and rolling back changes can be complicated:

Reviewer: This comment is still true; however, the new instructions are different from the old in now, not only are we making massive changes to the codebase, we are also going to switch our main branch to a completely different git history with no commits in common with the existing repository.

  git commit -m "Pull in configuration from default branch"
  ```

1. Check for `pantheon.yml` settings you need to preserve by comparing your old codebase's `pantheon.yml` to the new `pantheon.upstream.yml`:

  ```bash{promptUser:user}
  git diff master:pantheon.yml pantheon.upstream.yml
  ```

1. If there are settings from `pantheon.yml` (shown with a `-` in the diff output), consider copying over your old pantheon.yml to preserve these settings:

  ```bash{promptUser:user}
  git checkout master pantheon.yml
  git add pantheon.yml
  git commit -m 'Copy my pantheon.yml'
  ```

Note that a few settings must be left as they are in `pantheon.upstream.yml` for sites using Integrated Composer. These are:

* `web_docroot`
* `build_step`

### (Optional) Return to Drupal 8

If you’re not ready to update to Drupal 9 yet. (Advisable: test the Composer conversion before testing the Drupal 9 upgrade)
Note that you might want to manually edit `composer.json` and remove the `drupal/core` entry if it exists. The empty drupal/core entry from the steps above will cause drupal/core to be erased and redownloaded frequently

```bash{promptUser:user}
cd upstream-configuration/

> composer require drupal/core-recommended:^8 --no-update

Reviewer 4: Note 1: this will be inaccurate once https://github.com/pantheon-systems/drupal-project/pull/6 is released

Note 2: due to a new development / release process we're putting into effect, merging that PR is not the same as it being released. I'm making a note to ping @EdwardAngert when it's actually released.

  ```

  If you prefer, you can replace `composerify` with another branch name. If you do, remember to adjust the other examples in this doc to match.

1. Copy any existing configuration from the default branch. If no files are copied through this step, that's ok:

  ```bash{promptUser:user}
  git checkout master sites/default/config
  git mv sites/default/config/* config
  git rm -f sites/default/config/.htaccess
  git commit -m "Pull in configuration from default branch"
  ```

1. Check for `pantheon.yml` settings you need to preserve by comparing your old codebase's `pantheon.yml` to the new `pantheon.upstream.yml`:

  ```bash{promptUser:user}
  git diff master:pantheon.yml pantheon.upstream.yml
  ```

   - If there are settings from `pantheon.yml` (shown with a `-` in the diff output), consider copying over your old `pantheon.yml` to preserve these settings:

     ```bash{promptUser:user}
     git checkout master pantheon.yml
     git add pantheon.yml
     git commit -m 'Copy my pantheon.yml'
     ```

   If you prefer to keep the values for any of the following settings from `pantheon.upstream.yml`, remove them from `pantheon.yml`:

>      * `web_docroot`

Reviewer: `web_docroot` should always be inherited from the upstream.

Author: removed from this list and reframed the sentence to show `database` as the only value. Is that right?

     * `build_step`

>      * `database`

Reviewer: So right now the upstream tries to upgrade to 10.4, which isn't going to work for a while. Perhaps we need to instruct folks to set the database version to 10.0 until general availability.

Reviewer 4: I thought our approach to this was to release the documentation for performing conversions to IC only for now, staying on the major version of Drupal you're already on. The title says "Convert to Drupal 9", but we fundamentally can't do that, so we should change the title.

I thought we had an optional step describing what to do to stay on Drupal 8 and we would essentially just make it a required step. But I do not see it in the rendered documentation. Perhaps this is what I had that post-it I had to ping you was about @EdwardAngert -- the details of how to stay on Drupal 8 did change with the new release of the upstream. I'll take a stab at those steps now.

Reviewer 4: The comments are getting numerous and hard to follow, but it looks like I posted a recommended change for what makes sense to me at https://github.com/pantheon-systems/documentation/pull/6145#discussion_r575484057

Reviewer: The goal here is to put people on the pantheon-upstreams/drupal-project upstream. We could make an 8.x branch of that upstream, or make a different upstream if we want to simplify the instructions here, but if we keep on the current course, pantheon.upstream.yml sets the database version to 10.4.  It is therefore insufficient for us to remove the database version; we must reset it to 10.0 or the user will receive an error.

Should confirm -- I have not tried this for a while. My understanding, though, is that this error will persist until Mariadb / Drupal 9 general availability.

> When reviewing your site, take stock of exactly what versions of modules you depend on. One way to do this is to use a command like the following from within a contributed modules folder (e.g. `/modules`, etc):

Reviewer: Here you can say "modules and themes", and it's fine to just say "One way to do this is to use a command like the following:" (it doesn't matter which folder you're in when you run drush pm:projectinfo)

```bash{promptUser:user}

> terminus drush $site.dev -- pm:projectinfo --fields=name,version --format=table

Reviewer: Note that `pm:projectinfo` was removed in Drush 9, so the users will need to revert back to Drush 8 if they are using a newer Drush. Usually folks stay on Drush 8 until they convert to Composer, so this might not be a common problem.

Author: offered `pm:list` and linked to drush docs

```

This will list each module followed by the version of that module that is installed.

You can add these modules to your new codebase using Composer by running the following for each module in the `$site-composer` directory:

```bash{promptUser:user}

> composer require drupal/MODULE_NAME:^VERSION

Reviewer: It would probably be a good idea to tell folks that they should skip the ludwig module if they are using it. This module is used to integrate modules that require Composer into sites that are not composer-managed, so it might be fairly common.

Reviewer: Also, note that Drush will show version strings in two different format.

If Drush says 8.x-1.8, then they should use ^1.8 (common)

If Drush says 4.0.1, then they should use ^4.0.1 (less common)

Reviewer: Below, the section "#### Themes" may be removed, because it is now covered by the section above. Furthermore, the section "#### Libraries" is now largely unneeded. With Composer-managed sites, for the most part the libraries your modules will need will be included automatically by Composer. This will happen for any module that has been updated for Composer, which by now is the majority of them. There might be some modules remaining that still need external libraries, and do not use Composer to install them. These will need special handling. I'm not sure what to say in the "Libraries" section, though, since I don't have any specific examples where this is still needed.

Reviewer: The section "#### Modules and Themes" should also be adjusted:

> Custom code should be checked out from the master branch, which contains the original site:
> ```
> git checkout master modules/custom
> ```
> Note that the command above does not switch branches; it copies the contents of the `modules/custom` directory from the master branch onto the current branch.
>
> Foolow suit with any other custom code you may need to carry over from other directories.

Follow suit with any other custom code you need to carry over.

#### Settings.php

Your existing site may have customizations to `settings.php` or any other config files. Review these carefully and extract relevant changes from these files to copy over. Always review any file paths referenced in the code, as these paths may change in the transition to Composer.

It is not wise to completely overwrite the `settings.php` file with the old one, as there are customizations for moving the configuration directory you don't want to overwrite, as well as platform specific customizations.

```bash{promptUser:user}

> git checkout master sites/default/settings.php
> mv -f sites/default/settings.php web/sites/default/
> git rm -rf sites
> git diff

Reviewer 4: The irony here is that the below commands do what the above suggestion says is not wise. Here's an alternative incantation that's more in keeping with "extract relevant changes from these files"
```suggestion
# Ensure working tree is clean
git status
git checkout master
git checkout -b old-settings-php
git mv sites/default/settings.php web/sites/default
git commit -m 'Relocate settings.php'
git checkout composerify
# Interactively extract relevant hunks from old settings.php
git checkout -p old-settings-php web/sites/default/settings.php
```

Your existing site may have customizations to `settings.php` or any other config files. Review these carefully and extract relevant changes from these files to copy over. Always review any file paths referenced in the code, as these paths may change in the transition to Composer.

It is not wise to completely overwrite the `settings.php` file with the old one, as there are customizations for moving the configuration directory you don't want to overwrite, as well as platform specific customizations.

```bash{promptUser:user}
# Ensure working tree is clean
git status
git checkout master
git checkout -b old-settings-php
git mv sites/default/settings.php web/sites/default
git commit -m 'Relocate settings.php'
git checkout composerify
# Interactively extract relevant hunks from old settings.php
git checkout -p old-settings-php web/sites/default/settings.php

> ```

Reviewer: It's kind of a lot of steps just to get your settings.php in the right place on a separate branch (you also need a `mkdir -p web/sites/default` in the instructions above, as `web` does not exist on the original branch).

How about this instead:
```
git checkout master sites/default/settings.php
diff -Nup web/sites/default/settings.php sites/default/settings.php 
# Edit settings.php as needed
rm sites/default/settings.php
```
You don't get to use `git checkout -p`, but there are significantly fewer steps.

Author: can we resolve this conversation?

Reviewer: Actually we need one more command before the `rm`:
```
git restore --staged sites/default/settings.php 
```
I don't know if there's a better way to do this.

Reviewer: OK, here's a good solution:
```
git show master:sites/default/settings.php > web/sites/default/original-settings.php
diff -Nup --ignore-all-space web/sites/default/settings.php web/sites/default/original-settings.php
# edit web/sites/default/settings.php and commit as needed
rm web/sites/default/original-settings.php
```

Author: updated to match your latest comment, but I've kept the `git status` to catch people who're just running through steps and might have missed something staged
I can remove it if it's too in the way

Reviewer: `git status` is a good idea.

```bash{promptUser:user}
git push origin composerify && terminus env:create $site.dev composerify
```

This will set up the Multidev environment to receive and demo our changed code.

## Create a New Composer Project

1. In your local terminal, from the repository root of your Pantheon site, move a directory up:

  ```bash{promptUser:user}
  cd ..
  ```

2. Use Composer to create a new project, using the [Pantheon Drupal 8 Composer](https://github.com/pantheon-systems/example-drops-8-composer) repository:

    ```bash{promptUser:user}
    composer create-project pantheon-systems/example-drops-8-composer $site-composer
    cd $site-composer
    ```

This will create a new directory based on the example project [pantheon-systems/example-drops-8-composer](https://github.com/pantheon-systems/example-drops-8-composer) in the `$site-composer` directory.

## Update to Latest Drupal Core

The Pantheon Drupal 8 Composer repository may not depend on the latest secure release of Drupal Core. To keep your site safe, it's a good idea to pull the latest version in now -- before moving on -- by running:

```bash{promptUser:user}
composer update drupal/core --with-dependencies

> ```

Reviewer 4: These instructions are out of date -- suggest deleting this section.

## Prepare to Deploy

> At this point, your new project directory should contain all of the unique code from your existing Drupal 8 site, plus all of the code required to make a Composer driven project work. Since Pantheon requires all runtime code to be present when deploying to the platform, if no CI solution is a part of your workflow, you must now modify the project to be deployable straight to Pantheon.

Reviewer 4: Heads up to update this part about if no CI solution is available and using Build Tools. I assume we want to direct folks to Integrated Composer as the default now.

Locate the configuration files in your existing site and move them here. If they are stored in the files directory on your existing site, retrieve them via [SFTP](/sftp), as the Git clone would not contain them. The example project is configured to use this location.

## Set up a Multidev (Optional)

If your Pantheon account has access to [Multidev](/multidev), create a Multidev to push your new code to:

```bash{promptUser:user}

> git push origin composerify && terminus env:create $site.dev composerify

Reviewer: Here's the point where the intrepid adventurer may receive the error:

> Changing database version after environment creation is not currently supported.

If this happens, then remove the `database:` section from pantheon.yml. If the user does this, then they should also downgrade to Drupal 8 and upgrade to Drupal 9 later, after the database upgrade is available.

Reviewer: It would also probably be better to have the user create the 'composerify' multidev before pushing the branch. If you do it the other way, Pantheon won't necessarily pick up changes to the composer.json file.

```

This will set up the Multidev environment to receive and demo our changed code.

> ## Commit

Reviewer: This section ("Commit") may be removed.

If the user has been following along with the instructions in this document, then all of the work should have been done on the "composerify" branch, and at this point everything should be committed and pushed.

Commit your work to the Git repo. From the `$site` directory, run the following:

```bash{promptUser: user}
cp -r .git ../$site-composer/
cd ../$site-composer
git add .
git commit -m "Convert to Composer based install"
```

You should see a large number of files committed to the new branch created earlier.

> ## Prepare to Deploy

Reviewer: The "Prepare to Deploy" step should no longer be necessary. There is no `prepare-for-pantheon` any longer, and `composer install` should have already ran implicitly as one of the previous steps.

composer install --no-dev
```

This modifies the `.gitignore` file and cleans up any errant `.git` directories in the codebase to prepare your new code for direct deployment to Pantheon.

## Deploy

> You've now committed the code to the local branch. If your site has Multidev, you can deploy that branch directly to a new Multidev and test the site in the browser. If the site doesn't load properly, clear the cache. If there are any issues, utilize your site's logs via `terminus drush $site.composerify -- wd-show` to inspect the watchdog logs, or follow the directions in our documentation on [log collection](/logs).

Reviewer: This paragraph with `drush wd-show` could be moved up to the "Set up a Multidev" section above. The rest of this ("Deploy" and "Deploy to a Multidev") is redundant with what has come before.

### Deploy to a Multidev

If your site has Multidev, push the changes to a Multidev called `composerify` to safely test the site without affecting the Dev environment:

```bash{promptUser:user}
git add .
git commit -m "ran composer prepare-for-pantheon and install"
git push origin composerify
```

> Once you have confirmed the site is working, merge `composerify` into `master`, and follow the standard workflow to QA a code change before going live.

Reviewer: Maybe this sentence also should move up to the "Set up a Multidev" section.

### Deploy to Dev

> If you have a [local development](/local-development) solution, consider testing your `composerify` branch locally before merging.

Reviewer: Redundant with "Set up a Multidev" section (although you might add a note up there to skip that step if their plan does not support multidevs)

To force push the changes from the local branch to the `master` Dev branch:

```bash{promptUser:user}
git add .
git commit -m "ran composer prepare-for-pantheon and install"
git push -f origin composerify:master
```

## Change Upstreams

Set the site to use the Drupal 9 Upstream:

```bash{promptUser:user}

> terminus site:upstream:set $site drupal9

Reviewer: It should be safe to do this even if the user pinned to Drupal 8.

Author: should I just add the line `It should be safe to do this even if the site is pinned to Drupal 8.`? or is there more nuance here that I'm missing?

You should see a large amount of files committed to the new branch we created earlier.

## Deploy

You've now committed the code to a branch. If your site has Multidev, you can deploy that branch directly to a new Multidev and test the site in the browser. If the site doesn't load properly, clear the cache. If there are any issues, utilize your site's logs via `terminus drush $site.composerify -- wd-show` to inspect the watchdog logs, or follow the directions on our documentation on [log collection](/logs).

Once you have confirmed the site is working, merge `composerify` into `master`, and follow the standard workflow to QA a code change before going live.

If your plan does not include Multidev, you will have to merge to master before deploying, then follow the rest of the steps above. If you have a local development solution, consider testing your `composerify` branch locally before merging.

## Change Upstreams

Your Pantheon site is no longer compatible with traditional upstream updates. Avoid confusion by moving your site to an empty upstream:

```bash{promptUser:user}

> terminus site:upstream:set $site empty

Reviewer 4: I think we can now support -- and therefore want -- an end state where the site *is* compatible with traditional upstream updates. Therefore we want to point sites -- both Drupal 8 and 0 -- at the `drupal9` official upstream. That way these sites will benefit from receiving all updates through the normal flow.
```suggestion
terminus site:upstream:set $site drupal9
```

```

> ## Ongoing Core Updates

Reviewer 4: Heads up to update this part to recommend use of the dashboard's check for updates / apply updates functionality when using Integrated Composer

## Deploy

You've now committed the code to the local branch. Deploy that branch directly to a new Multidev and test the site in the browser. If the site doesn't load properly, clear the cache. If there are any issues, utilize the site's logs via `terminus drush $SITE.composerify -- wd-show` to inspect the watchdog logs, or follow the directions in our documentation on [log collection](/logs).

### Deploy to a Multidev

Push the changes to a Multidev called `composerify` to safely test the site without affecting the Dev environment:

```bash{promptUser:user}
git add .
git commit -m "ran composer prepare-for-pantheon and install"
git push origin composerify && terminus env:create $SITE.dev composerify
```

Once you have confirmed the site is working, merge `composerify` into `master`, and follow the standard workflow to QA a code change before going live.

### Deploy to Dev

If you have a [local development](/local-development) solution, consider testing your `composerify` branch locally before merging.

To force push the changes from the local branch to the `master` Dev branch:

```bash{promptUser:user}
git add .

> git commit -m "ran composer prepare-for-pantheon and install"

Reviewer: Comment does not match instructions; `composer prepare-for-pantheon` isn't part of this guide (this is from the guild on converting a site to build tools)
